File: DFS/133.clone-graph.go
package leetcode

// @leet start
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
	if node == nil {
		return nil
	}

	copies := [101]*Node{}
	var dfs func(node *Node)
	dfs = func(node *Node) {
		if node == nil { // not necessary. since the range will exclude the nil node
			// eventually
			return
		}
		newNode := &Node{Val: node.Val}
		copies[node.Val] = newNode

		for _, neighbor := range node.Neighbors {
			if copies[neighbor.Val] == nil {
				dfs(neighbor)
			}
			newNode.Neighbors = append(newNode.Neighbors, copies[neighbor.Val])
		}
	}
	dfs(node)
	return copies[node.Val]
}

// @leet end
func cloneGraph(node *Node) *Node {
	var (
		copied = map[*Node]*Node{}
		dfs    func(*Node)
	)
	dfs = func(node *Node) {
		if node == nil {
			return
		}
		copied[node] = &Node{Val: node.Val}
		for _, item := range node.Neighbors {
			if copied[item] == nil {
				dfs(item)
			}
			copied[node].Neighbors = append(copied[node].Neighbors, copied[item])
		}
	}
	dfs(node)
	return copied[node]
}

// @leet end

func cloneGraph2(node *Node) *Node {
	if node == nil {
		return nil
	}
	nn := &Node{Val: node.Val}
	for _, v := range node.Neighbors {
		nn.Neighbors = append(nn.Neighbors, cloneGraph(v))
		// NOTE: this will not work seem it will self connected.
	}
	return nn
}

File: DFS/133.clone-graph.py
from type import Node
from typing import Optional
from collections import deque

# @leet start
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        m = {}  # NOTE: simple map

        def dfs(node):
            if node is None:
                return
            m[node] = Node(node.val)
            for item in node.neighbors:
                if item not in m:
                    dfs(item)
                m[node].neighbors.append(m[item])

        dfs(node)
        if node in m:
            return m[node]
        return None




class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if node is None:
            return None
        old_new: dict[Node, Node] = {}
        old_new[node] = Node(node.val)
        q: deque[Node] = deque([node])
        while len(q) > 0:
            cur = q.popleft()
            cur_copy = old_new[cur]
            for item in cur.neighbors:
                if item not in old_new:
                    q.append(item)  # NOTE: to append. the order may be different
                    nn = Node(item.val)
                    old_new[item] = nn
                copy_item = old_new[item]
                cur_copy.neighbors.append(copy_item)

        return old_new[node]


# @leet end


class Solution1:  # NOTE: self connected. timeout
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if node is None:
            return None
        nn = Node(node.val)
        for item in node.neighbors:
            nn.neighbors.append(self.cloneGraph(item))
        return nn


class Solution2:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if node is None:
            return None
        copyList: list[None | Node] = [None] * 101
        # head = Node(node.val)
        # copyList[head.val] = head

        def dfs(n: Node):
            # if n is None:
            #     return
            nn = Node(n.val)
            copyList[n.val] = nn
            for item in n.neighbors:  # NOTE: item will not be nil!!
                # neighber can be empy list
                if copyList[item.val] is None:
                    dfs(item)
                nn.neighbors.append(copyList[item.val])  # type: ignore

        dfs(node)
        return copyList[node.val]
        # return head  # NOTE: note work first try will overwrite the head instance
